packages/gatsby-transformer-mdx-introspection/src/reduce-jsx-ast.js

```diff
@@ -0,0 +1,307 @@
+const {
+  reduceNode,
+  collapseSpace,
+  isJsxElement,
+  getJsxChildren,
+} = require('./jsx-ast-utils');
+
+// Contains methods that form a transformation pipeline converting the JSX
+// AST outputted by Babel's parser to a simpler, reduced component tree that
+// is used as the data model for the plugin
+
+/**
+ * Serializes a node to string and searches for any embedded lingering JSX
+ * tags contained within, adding them to the detachedHeads array for further
+ * processing as a separate component tree
+ * @param {object} node Babel JSX AST
+ * @param {boolean} collapse Whether to collapse whitespace in JSX snippets
+ * @param {array} detachedHeads Mutable array of detached head Babel JSX AST
+ * nodes
+ * @param {object} context Context object
+ */
+function serializeAndSearch(node, collapse, detachedHeads, context) {
+  const { cleanWhitespace, jsx } = context;
+  const value = reduceNode(node, cleanWhitespace && collapse, jsx);
+  const subJsx = getJsxChildren(node);
+  if (subJsx.length > 0) {
+    // The value is (or contains) one or more JSX tree heads; add to list
+    detachedHeads.push(...subJsx);
```

Comment:
Not sure I like this side effect. I would prefer to pass a callback function instead and let the consumer do the array.push outside of this method.

```js
function serializeAndSearch(node, collapse, onJsxChildren, context) {
  // ...
  if (subJsx.length > 0) {
    onJsxChildren(subJsx);
  }
  return value;
}
```

Reply:
I feel like the mutable array parameter isn't great as well-- it never felt like a good solution. I adapted it to use a callback to remove the mutable parameter.